File: 0202-happy-number/0202-happy-number.kt
class Solution {

    private val set = HashSet<Int>()

    fun isHappy(n: Int): Boolean {
        var sum = 0
        n.toString().forEach { num ->
            sum += num.toString().toInt() * num.toString().toInt()
        }
        if (sum == 1) {
            return true
        }
        if (set.contains(sum)) {
            return false
        }
        set.add(sum)
        return isHappy(sum)
    }
}

File: 0202-happy-number/0202-happy-number.swift
class Solution {
    
    private func getSquare(_ n: Int) -> Int {
        var result = 0
        var n = n
        while n != 0 {
            let remainder = n % 10
            n -= remainder
            n /= 10
            
            result += remainder * remainder
        }
        return result
        
    }
    
    func isHappy(_ n: Int) -> Bool {
        var memoi: Set<Int> = []
        var curr = n
        
        repeat {
            if memoi.contains(curr) {
                return false
            }
            memoi.insert(curr)
            curr = getSquare(curr)
        } while curr != 1
        
        return true
    }
}
